### ISSUE TITLE
Add an optional "at-fork" strategy

### ISSUE THREAD

**Issue by Snaipe** [nanomsg/nanomsg#582]
I was talking about this on the gitter channel with @gdamore, so I thought I'd write up my thoughts on the matter here.

(Sorry for the long wall of text, see the end for a tl;dr)
#### The context

I am maintaining an unit testing framework for C. One of its feature is that it runs each test in its separate address space, to make sure that in the not-so-unlikely event that the user code crashes, the test runner will be still alive and well.

Currently, I am gathering data from the workers through a pipe I set up. This was simple enough until now, although I had to make a portable abstraction layer over `pipe(2)` and `CreatePipe` to have it run on both posix & windows systems -- hence, the current structure of the runner is to create the pipe, then fork the workers, read data from the pipe, and fork new workers as old ones die.

While this was sufficient, it was a bit of a hassle to maintain and very limited for new purposes like having external workers (e.g. in an embedded device), along with a lot of other problems; this is why I considered using nanomsg for the rewrite of the IO layer.
#### The problem

Reading through the sources of nanomsg, I noticed that there was heavy usage of non-blocking I/O, and hence threads. This is usually a problem for programs using fork(), as the two are orthogonal to each other, so I went to see if there was a `pthread_atfork` handle.

It turns out there isn't (I don't blame you, this is an issue that cannot be trivially solved), which brings up the main topic of this issue: what should be done on the side of nanomsg if a process decides to fork() ?

I agree that there cannot be an "expected" behaviour and that it depends on the use case, which is
why I think nanomsg could have a function to set an atfork strategy before the first socket is ever opened.
#### The current stategy

Currently, this is what happens when forking:
1. The parent calls fork()
2. The child process spawns, all (possibly locked) mutexes and file descriptors are inherited from the parent.
3. The parent resumes.

This means that nanomsg is effectively in an unusable state in the child process after the fork.

Thankfully, nanomsg also sets `FD_CLOEXEC` on its sockets, so for the common fork()-then-exec() use case it's possible to reset the library to an usable state.
#### A possible new strategy

In my case, I don't call exec() after fork, because I still want to work with a clone of the current process. However, it turns out that I don't really need any of the sockets from the parent in the child process: I could easily connect in the child to the server socket in the parent. I would just need nanomsg to be in a "clean" state, as if no sockets existed.

Effectively, this would be equivalent to manually do what happens on exec(), but earlier: release all the locks and close all of nanomsg's file descriptors.

A naive implementation (in pseudo-code, as I am not familiar enough with the internals of nanomsg) could be:

``` c
void nn_prefork(void) {
    // for each worker thread
    for (struct nn_worker *w = ...; ...; ...) {
        // Lock its mutex
        nn_mutex_lock (&w->sync);
    }

    // lock the global lock
    nn_glock_lock ();
}

void nn_postfork_parent(void) {
    // unlock the global lock
    nn_glock_unlock ();

    // for each worker thread
    for (struct nn_worker *w = ...; ...; ...) {
        // Unlock its mutex
        nn_mutex_unlock (&w->sync);
    }
}

void nn_postfork_child(void) {
    // unlock the global lock
    nn_glock_unlock ();

    // for each worker thread
    for (struct nn_worker *w = ...; ...; ...) {
        // Unlock its mutex
        nn_mutex_unlock (&w->sync);

        // Clean up worker resources
        nn_timerset_term (&w->timerset);
        nn_poller_term (&w->poller);
        nn_efd_term (&w->efd);
        nn_queue_item_term (&w->stop);
        nn_queue_term (&w->tasks);
        nn_mutex_term (&w->sync);
    }

    // for each opened socket
    for (int *sock = ...; ...; ...) {
        nn_close(*sock);
    }
}

enum nn_option {
    NN_FORK_STRATEGY,
};

enum nn_fork_strategy {
    NN_FORK_NONE,
    NN_FORK_RESET
};

int nn_setopt (int option, const void *optval, size_t optvallen) {
    ...
    if (option == NN_FORK_STRATEGY && ((const int *)optval) == NN_FORK_RESET) {
        pthread_atfork(nn_prefork, nn_postfork_parent, nn_postfork_child);
    }
    ...
}
```

---

Maybe another "cleaner" possibility:

``` c
void nn_prefork(void) {
    // for each worker thread
    for (struct nn_worker *w = ...; ...; ...) {
        // Ask worker to suspend itself
        nn_mutex_lock (&w->sync);
        nn_queue_push (&w->tasks, &w->suspend);
        nn_efd_signal (&w->efd);
        nn_mutex_unlock (&w->sync);

        // Wait until the worker is suspended
        nn_wait_thread_suspended (&self->thread); // TODO

        // Reacquire the lock
        nn_mutex_lock (&w->sync);
    }
}

void nn_postfork_parent(void) {
    // for each worker thread
    for (struct nn_worker *w = ...; ...; ...) {
        // Wake worker up
        nn_thread_wakeup (w) // TODO

        // Unlock its mutex
        nn_mutex_unlock (&w->sync);
    }

    // lock the global lock
    nn_glock_lock ();
}

// Same nn_postfork_child and nn_setopt.
```

And modify the current posix worker to lock the sync mutex and wait on a `pthread_cond_t` or something like this.

Since waiting on a `pthread_cond_t` unlocks its associated mutex, one can lock again the mutex, fork, and in the parent tell the thread to wake up (which will cause the waiting on the `pthread_cond_t` to try and lock the mutex), then unlock the mutex to allow the worker to resume.
#### TL;DR
- There is no current atfork strategy, which means that it's impossible to use nanomsg in a child process after fork()ing
- The sockets are set to close on exec(), so the only way to reset the library to an usable state is by calling exec() after fork() or vfork().
- Rather than trying to solve the problem, the library could provide functions in its API to let the user set the atfork strategy, based on his use case.
- One of the possible strategies could be to do the whole library reset -- that is normally done by the OS on exec() -- right after fork().

**Comment by gdamore** [comment]
I think nn_term in the child is probably the best approach.  That said, I really don't see a perfect way forward here.

**Comment by gdamore** [comment]
The other thing is that fork() can occur at any time, so we'd have to be sure that we *can* acquire the locks -- all of them.  I'm concerned of a potential for deadlock here.  Probably not a problem, since we don't fork from inside the library...

**Comment by Snaipe** [comment]
> I think nn_term in the child is probably the best approach.

Is the behaviour of nn_term well defined when the underlying thread is already dead? It might be better to simply clean up the resources rather than trying to notify and join the dead thread.

> The other thing is that fork() can occur at any time, so we'd have to be sure that we can acquire the locks -- all of them. I'm concerned of a potential for deadlock here. Probably not a problem, since we don't fork from inside the library...

This shouldn't be a problem indeed. The atfork handler is run on the thread calling fork(), and all the mutex locks/unlocks are done on threads owned by nanomsg.

The only issue I see would be to have a thread doing some blocking I/O while owning a lock. -- this would make fork() an excruciatingly slow operation -- however, I am not familiar enough with the internals to know that, and a quick look through `src/aio/worker_posix.inc` seems to indicate that it wouldn't happen.

**Comment by reqshark** [comment]
> Is the behaviour of nn_term well defined when the underlying threads are already dead (or rather 'not existing') ? It might be better to simply clean up the resources rather than trying to notify and join the dead threads.

that's exactly what it does, it's like `nn_close()` on steroids

> The only issue I see would be to have a thread doing some blocking I/O while owning a lock. --  this would make fork() an excruciatingly slow operation

interesting notion, i never thought of that particular angle, but I knew fork() was slow.

OK my opinion on the subject... having a set of initialized processes running for task absorption before hand is one thing.. they're already setup. but needlessly making high latency calls to fork and exec because it's already too late... that's something different. there's a prohibitive amount of upfront startup latency associated with new process or fork spinup.. so do it early if u have to

and don't do it again during end-user application experience lifecycles

**Comment by gdamore** [comment]
Having spent a bunch of time researching this, we cannot actually do anything legal and useful during pthread_atfork() because in the child we are not permitted to use async-signal-unsafe functions.  This means that the child side cannot use mutexes, and cannot do many useful things. (It could "close" some descriptors, but that's about the limit.)

To throw some more fuel on this fire, POSIX is basically abandoning this strategy, because its too hard to get right.  And, interestingly enough, posix_spawn() does not call these handlers for some platforms, so we can't even block things to close the close-on-exec flag races.

In the future we may actually go out of our way to do nn_panic() if you try to reenter the library from a fork'd child (that hasn't been reset via exec).

Fundamentally, we are not going to be fork-safe now, or ever. Sorry.
